src/types/string.rs

```diff
@@ -266,29 +268,30 @@ impl ZendStr {
     /// Returns a reference to the underlying [`CStr`] inside the Zend string.
     pub fn as_c_str(&self) -> &CStr {
         // SAFETY: Zend strings store their readable length in a fat pointer.
-        unsafe {
-            let slice = slice::from_raw_parts(self.val.as_ptr() as *const u8, self.len() + 1);
-            CStr::from_bytes_with_nul_unchecked(slice)
-        }
+        unsafe { CStr::from_bytes_with_nul_unchecked(self.as_bytes()) }
     }
 
-    /// Attempts to return a reference to the underlying [`str`] inside the Zend
+    /// Attempts to return a reference to the underlying bytes inside the Zend
     /// string.
     ///
-    /// Returns the [`None`] variant if the [`CStr`] contains non-UTF-8
-    /// characters.
+    /// Returns an [Error::InvalidUtf8] variant if the [`str`] contains
+    /// non-UTF-8 characters.
     ///
     /// # Example
     ///
     /// ```no_run
     /// use ext_php_rs::types::ZendStr;
     ///
-    /// let s = ZendStr::new("hello, world!", false).unwrap();
-    /// let as_str = s.as_str();
-    /// assert_eq!(as_str, Some("hello, world!"));
+    /// let s = ZendStr::new("hello, world!", false);
+    /// assert!(s.as_str().is_ok());
     /// ```
-    pub fn as_str(&self) -> Option<&str> {
-        self.as_c_str().to_str().ok()
+    pub fn as_str(&self) -> Result<&str> {
+        std::str::from_utf8(self.as_bytes()).map_err(|_| Error::InvalidUtf8)
```

Comment:
I'm wary that we might have a Schlemiel the Painter problem here.
If I remember correctly, there's a flag somewhere at the C-level to indicate that the string has already been utf8-validated (i.e. by the mbstring or pcre extensions), need to investigate...

Reply:
It's [here](https://github.com/php/php-src/blob/master/ext/mbstring/mbstring.c#L4369) and also [here](https://github.com/php/php-src/blob/master/ext/pcre/php_pcre.c#L1469).
So we could do something like:
```rs
if GC_FLAGS(self) & IS_STR_VALID_UTF8 {
   return std::str::from_utf8_unchecked(...);
}
let s = std::str::from_utf8(self.as_bytes()).map_err(|_| Error::InvalidUtf8)?;
GC_ADD_FLAG(self, IS_STR_VALID_UTF8):
return s;
```

Reply:
This requires adding stuff to the ffi bindings though (the gc flags related stuff in `zend_types.h`).
I'm going to leave this off for now and do it in a separate PR.

Reply:
Dors this require mbstring to be installed

Reply:
> Dors this require mbstring to be installed

No, the flag is _used_ by `mbstring` and `PCRE` but it is defined in [zend_types.h](https://github.com/php/php-src/blob/eb83e0206c9b9261d786943bf2c5ad61dca287e2/Zend/zend_types.h#L721)

Reply:
Yes I think so too!
I proposed to do this in another PR to keep this one focused, but if you think it should be done here I'm OK with that.
Do you confirm?

Reply:
Yes, this was my intention. Sorry and thank you

Reply:
It shall be done then! 😉